File: FSharpForCSharpDevelopers/FSharpApplication/DecoratorPattern.fs
module temp

let addingCalculator input = input + 1
let multiplyingCalc input = input * 5
let subtractingCalculator x y = x -y

let loggingCalculator innerCalculator input = 
   printfn "input is %A" input
   let result = innerCalculator input
   printfn "result is %A" result
   result

let fileLoggingCalculator innerCalculator input = 
   printfn "input is %A" input
   let result = innerCalculator input
   printfn "result is %A" result
   result


let result = loggingCalculator addingCalculator 5

let composedFunc = loggingCalculator addingCalculator

let result2 = fileLoggingCalculator multiplyingCalc 10

let result10 = composedFunc 10

let result5 = composedFunc 5

File: FSharpForCSharpDevelopers/CSharpApplication/DecoratorPattern.cs
using System;

namespace CSharpApplication
{
    interface ICalculator
    {
        int Calculate(int input);
    }

    class AddingCalculator : ICalculator
    {
        public int Calculate(int input) { return input + 1; }
    }

    class LoggingCalculator : ICalculator
    {
        readonly ICalculator _innerCalculator;

        LoggingCalculator(ICalculator innerCalculator)
        {
            _innerCalculator = innerCalculator;
        }

        public int Calculate(int input)
        {
            Console.WriteLine("input is {0}", input);
            var result = _innerCalculator.Calculate(input);
            Console.WriteLine("result is {0}", result);
            return result;
        }
    }
}